Mathlib/Analysis/Asymptotics/LinearGrowth.lean

```diff
@@ -0,0 +1,616 @@
+lemma eventually_le_mul (h : linearGrowthSup u < a) :
+    ∀ᶠ n : ℕ in atTop, u n ≤ a * n :=
+  linearGrowthSup_le_iff.1 (le_refl (linearGrowthSup u)) a h
+
+/- Forward direction of `le_linearGrowthSup_iff`. -/
+lemma frequently_mul_le (h : a < linearGrowthSup u) :
+    ∃ᶠ n : ℕ in atTop, a * n ≤ u n :=
+  le_linearGrowthSup_iff.1 (le_refl (linearGrowthSup u)) a h
+
+lemma _root_.Frequently.linearGrowthInf_le (h : ∃ᶠ n : ℕ in atTop, u n ≤ a * n) :
+    linearGrowthInf u ≤ a := by
+  apply linearGrowthInf_le_iff.2 fun c c_u ↦ h.mono fun n hn ↦ hn.trans ?_
+  exact mul_le_mul_of_nonneg_right c_u.le n.cast_nonneg'
+
+lemma _root_.Eventually.le_linearGrowthInf (h : ∀ᶠ n : ℕ in atTop, a * n ≤ u n) :
+    a ≤ linearGrowthInf u := by
+  apply le_linearGrowthInf_iff.2 fun c c_u ↦ h.mono fun n hn ↦ hn.trans' ?_
+  exact mul_le_mul_of_nonneg_right c_u.le n.cast_nonneg'
+
+lemma _root_.Eventually.linearGrowthSup_le (h : ∀ᶠ n : ℕ in atTop, u n ≤ a * n) :
+    linearGrowthSup u ≤ a:= by
+  apply linearGrowthSup_le_iff.2 fun c c_u ↦ h.mono fun n hn ↦ hn.trans ?_
+  exact mul_le_mul_of_nonneg_right c_u.le n.cast_nonneg'
+
+lemma _root_.Frequently.le_linearGrowthSup (h : ∃ᶠ n : ℕ in atTop, a * n ≤ u n) :
+    a ≤ linearGrowthSup u := by
+  apply le_linearGrowthSup_iff.2 fun c c_u ↦ h.mono fun n hn ↦ hn.trans' ?_
+  exact mul_le_mul_of_nonneg_right c_u.le n.cast_nonneg'
+
+/-! ### Special cases -/
+
+lemma linearGrowthSup_bot : linearGrowthSup (⊥ : ℕ → EReal) = (⊥ : EReal) := by
+  nth_rw 2 [← limsup_const (f := atTop (α := ℕ)) ⊥]
+  refine limsup_congr (eventually_atTop.2 ?_)
+  exact ⟨1, fun n n_pos ↦ bot_div_of_pos_ne_top (Nat.cast_pos'.2 n_pos) (natCast_ne_top n)⟩
+
+lemma linearGrowthInf_bot : linearGrowthInf (⊥ : ℕ → EReal) = (⊥ : EReal) := by
+  apply le_bot_iff.1
+  rw [← linearGrowthSup_bot]
+  exact linearGrowthInf_le_linearGrowthSup
+
+  lemma linearGrowthInf_top : linearGrowthInf ⊤ = (⊤ : EReal) := by
+  nth_rw 2 [← liminf_const (f := atTop (α := ℕ)) ⊤]
+  refine liminf_congr (eventually_atTop.2 ?_)
+  exact ⟨1, fun n n_pos ↦ top_div_of_pos_ne_top (Nat.cast_pos'.2 n_pos) (natCast_ne_top n)⟩
+
+lemma linearGrowthSup_top : linearGrowthSup (⊤ : ℕ → EReal) = (⊤ : EReal) := by
+  apply top_le_iff.1
+  rw [← linearGrowthInf_top]
+  exact linearGrowthInf_le_linearGrowthSup
+
+lemma linearGrowthInf_const (h : b ≠ ⊥) (h' : b ≠ ⊤) : linearGrowthInf (fun _ ↦ b) = 0 :=
+  (tendsto_const_div_atTop_nhds_zero_nat h h').liminf_eq
+
+lemma linearGrowthSup_const (h : b ≠ ⊥) (h' : b ≠ ⊤) : linearGrowthSup (fun _ ↦ b) = 0 :=
+  (tendsto_const_div_atTop_nhds_zero_nat h h').limsup_eq
+
+lemma linearGrowthInf_zero : linearGrowthInf 0 = (0 : EReal) := by
+  rw [Pi.zero_def]
+  exact linearGrowthInf_const zero_ne_bot zero_ne_top
+
+lemma linearGrowthSup_zero : linearGrowthSup 0 = (0 : EReal) := by
+  rw [Pi.zero_def]
+  exact linearGrowthSup_const zero_ne_bot zero_ne_top
+
+lemma linearGrowthInf_mul : linearGrowthInf (fun n ↦ a * n) = a :=
```

Comment:
```suggestion
lemma linearGrowthInf_const_mul_self : linearGrowthInf (fun n ↦ a * n) = a :=
```
I'd use a more specific name because the statement is quite specific: I would expect a generic statement like `linearGrowthInf_mul` to talk about the behavior under multiplication of the growth of two general sequences.

Same thing below.